### ISSUE TITLE
Log file on Windows contains progress status text, escape characters and extra line breaks

### ISSUE THREAD

**Issue by uok** [restic/restic#3111]
Hello,

The log file on Windows contains progress status text, escape characters and extra line breaks. 
This does not happen on Linux. (https://forum.restic.net/t/different-log-file-in-windows-and-linux/3273)

Please fix, thanks!

Output of `restic version`
--------------------------

Windows
`restic 0.11.0 compiled with go1.15.3 on windows/amd64`
Linux
`restic 0.11.0 compiled with go1.15.3 on linux/amd64`

How did you run restic exactly?
-------------------------------

Windows
`restic backup --repo C:\Backup C:\Users --cleanup-cache --verbose --verbose >> C:\restic.log 2>&1`
Linux
`restic backup --repo /backup /source --cleanup-cache --verbose --verbose >> /restic.log 2>&1`

Expected behavior
-----------------

log files are the same on Windows and Linux

Linux

```
open repository
lock repository
load index files
using parent snapshot 5f9668c4
start scan on [/source]
start backup on [/source]
new       /source/file.txt, saved in 0.044s (100 B added)
unchanged /source/file.jpg
...
```

Actual behavior
---------------

Windows

```
open repository

e[2Klock repository

e[2Kload index files

e[2Kusing parent snapshot cf8362f4

e[2Kstart scan on [C:\Users]

e[2Kstart backup on [C:\Users]

e[2K[0:00] 0 files, 0 B, 0 errors

e[2Kunchanged /C/Users/MyUser/Config.xml

e[2K[0:00] 0 files, 0 B, 0 errors

e[2Kunchanged /C/Users/MyUser/test.html
...
```

(e[ = Escape character)

Do you have an idea how to solve the issue?
-------------------------------------------

This did not happen in older versions <= 0.9.5

Did restic help you today? Did it make you happy in any way?
------------------------------------------------------------

I'm happy that Restic also works on Windows

**Comment by MichaelEischer** [comment]
@uok Did you test whether restic 0.9.6 and/or 0.10.0 also print escape characters? Which terminal do you use? cmd, powershell or something else? Apparently restic considers the output file to be a terminal.

**Comment by dimejo** [comment]
I'm seeing the same behaviour as @uok when running restic v0.10.0 or newer from a batch file or directly in cmd. Powershell outputs looks fine though.

restic v0.9.6:
```
created new cache in C:\Users\aj\AppData\Local\restic

Files:          50 new,     0 changed,     0 unmodified
Dirs:            4 new,     0 changed,     0 unmodified
Added to the repo: 105.673 MiB

processed 50 files, 106.260 MiB in 0:02
snapshot b4c894fb saved
```
restic v.0.10.0:

```
[2K[0:00] 0 files 0 B, total 1 files 782.249 KiB, 0 errors

[2K[0:00] 8 files 7.771 MiB, total 100 files 74.230 MiB, 0 errors

[2K/D/test/SCENE009.JPG

[2K/D/test/SCENE010.JPG
[1A[1A
[2K[0:00] 9 files 8.634 MiB, total 100 files 74.230 MiB, 0 errors

[2K/D/test/SCENE009.JPG

[2K/D/test/SCENE010.JPG
[1A[1A
[2K[0:00] 15 files 13.743 MiB, total 100 files 74.230 MiB, 0 errors

[2K/D/test/SCENE016.JPG

[2K/D/test/SCENE017.JPG
[1A[1A
[2K[0:00] 20 files 17.216 MiB, total 100 files 74.230 MiB, 0 errors

[2K/D/test/SCENE016.JPG

[2K/D/test/SCENE021.JPG
[1A[1A
[2K[0:00] 26 files 21.516 MiB, total 100 files 74.230 MiB, 0 errors

[2K/D/test/SCENE021.JPG

[2K/D/test/SCENE027.JPG
[1A[1A
[2K[0:00] 33 files 25.886 MiB, total 100 files 74.230 MiB, 0 errors

[2K/D/test/SCENE027.JPG

[2K/D/test/SCENE034.JPG
[1A[1A
[2K[0:00] 37 files 29.984 MiB, total 100 files 74.230 MiB, 0 errors

[2K/D/test/SCENE034.JPG

[2K/D/test/SCENE038.JPG
[1A[1A
[2K[0:00] 44 files 34.121 MiB, total 100 files 74.230 MiB, 0 errors

[2K/D/test/SCENE038.JPG

[2K/D/test/SCENE045.JPG
[1A[1A
[2K[0:01] 49 files 38.220 MiB, total 100 files 74.230 MiB, 0 errors

[2K/D/test/SCENE045.JPG

[2K/D/test/SCENE051.JPG
[1A[1A
[2K[0:01] 50 files 38.220 MiB, total 100 files 74.230 MiB, 0 errors

[2K/D/test/SCENE045.JPG

[2K/D/test/SCENE051.JPG
[1A[1A
[2K[0:01] 56 files 42.642 MiB, total 100 files 74.230 MiB, 0 errors

[2K/D/test/SCENE051.JPG

[2K/D/test/SCENE057.JPG
[1A[1A
[2K[0:01] 63 files 48.214 MiB, total 100 files 74.230 MiB, 0 errors

[2K/D/test/SCENE057.JPG

[2K/D/test/SCENE064.JPG
[1A[1A
[2K[0:01] 66 files 50.224 MiB, total 100 files 74.230 MiB, 0 errors

[2K/D/test/SCENE057.JPG

[2K/D/test/SCENE067.JPG
[1A[1A
[2K[0:01] 69 files 53.378 MiB, total 100 files 74.230 MiB, 0 errors

[2K/D/test/SCENE067.JPG

[2K/D/test/SCENE070.JPG
[1A[1A
[2K[0:01] 74 files 56.468 MiB, total 100 files 74.230 MiB, 0 errors

[2K/D/test/SCENE070.JPG

[2K/D/test/SCENE075.JPG
[1A[1A
[2K[0:01] 78 files 60.225 MiB, total 100 files 74.230 MiB, 0 errors

[2K/D/test/SCENE075.JPG

[2K/D/test/SCENE079.JPG
[1A[1A
[2K[0:01] 86 files 64.516 MiB, total 100 files 74.230 MiB, 0 errors

[2K/D/test/SCENE079.JPG

[2K/D/test/SCENE087.JPG
[1A[1A
[2K[0:01] 92 files 68.789 MiB, total 100 files 74.230 MiB, 0 errors

[2K/D/test/SCENE087.JPG

[2K/D/test/SCENE093.JPG
[1A[1A
[2K[0:01] 98 files 73.067 MiB, total 100 files 74.230 MiB, 0 errors

[2K/D/test/SCENE093.JPG

[2K/D/test/SCENE099.JPG
[1A[1A
[2K[0:01] 100 files 74.230 MiB, total 100 files 74.230 MiB, 0 errors

[2K/D/test/SCENE099.JPG
[1A
[2K[0:02] 100 files 74.230 MiB, total 100 files 74.230 MiB, 0 errors

[2K

[2K[0:02] 100 files 74.230 MiB, total 100 files 74.230 MiB, 0 errors

[2KFiles:         100 new,     0 changed,     0 unmodified

[2K[0:02] 100 files 74.230 MiB, total 100 files 74.230 MiB, 0 errors

[2KDirs:            5 new,     0 changed,     0 unmodified

[2K[0:02] 100 files 74.230 MiB, total 100 files 74.230 MiB, 0 errors

[2KAdded to the repo: 74.259 MiB

[2K[0:02] 100 files 74.230 MiB, total 100 files 74.230 MiB, 0 errors

[2K

[2K[0:02] 100 files 74.230 MiB, total 100 files 74.230 MiB, 0 errors

[2Kprocessed 100 files, 74.230 MiB in 0:02

[2K[0:02] 100 files 74.230 MiB, total 100 files 74.230 MiB, 0 errors

[2Ksnapshot 06ecdc1d saved

[2K[0:02] 100 files 74.230 MiB, total 100 files 74.230 MiB, 0 errors

[2K

[1A
```

**Comment by uok** [comment]
@MichaelEischer, I'm using cmd directly or call restic from inside batch file - with same results.

**Comment by uok** [comment]
Unfortunately v0.12.0 still outputs the escape characters

**Comment by fd0** [comment]
Hm, interesting. Can you please try the latest beta from here: https://beta.restic.net/restic-v0.12.0-4-g8eb6a580/ and report back? Maybe this was https://github.com/restic/restic/pull/3286

**Comment by uok** [comment]
Thanks for your suggestion but this beta gives same result

**Comment by wiserain** [comment]
I am also experiencing this issue as @dimejo described. The only difference is that it appears to me after upgrading restic to v0.12. I am pretty sure it was okay before, though I don't remember which version I was with.

**Comment by fd0** [comment]
I'm reopening this issue, somebody needs to step up and look into it. Thanks for the reports!

**Comment by MichaelEischer** [comment]
@uok I cannot reproduce the issue when using cmd.
@wiserain I assume you're using powershell? Which restic version did you use before?

The problem is caused by the special case for mintty/cygwin at https://github.com/restic/restic/blob/8eb6a5805bdf290bee7fb4af1ace60a52a66197b/internal/ui/termstatus/terminal_windows.go#L92-L94 . Apparently powershell does not directly connect restic to the output file but uses a pipe instead. Commenting out the special case solves the problem for me.

I'm not sure whether there's a reliable indication whether restic is run by mintty?

**Comment by wiserain** [comment]
@MichaelEischer I am not sure, but I believe it was v0.11 cause I set my backup and got notified since 29 Dec last year. Note that v0.11 tagged on Nov 5 2020.

Yes, the result is obtained with powershell and pushed via discord webhook for notification. Please find my code and its corresponding output later.

**Comment by uok** [comment]
I use a batch file, maybe this is different? To keep only new/modified entries, I use `findstr` to filter unwanted lines
`restic backup --repo C:\Backup C:\Users --verbose --verbose | findstr /v /c:"unchanged /" | findstr /v /c:"scan finished" >> log.txt`
This works as expected in v0.11.0 but not in v0.12.0

**Comment by shd2h** [comment]
I can also confirm the issue shows up for me in powershell after updating to v0.12.0, and reverting to v0.11.0 fixes the issue.

I did a bit of testing; as far as I can tell, the issue _currently_ seems to only affect powershell, and only when redirecting restic output to a log file. If the output isn't redirected, there's no problem. I checked with both Tee-Object, Out-File and normal redirection as below, both in a simple powershell script and just in a live terminal:

```
D:\restic\restic.exe --repo 'C:\restic_repo' --password-file 'C:\repopass' --verbose backup 'C:\Morrowind' | Out-File 'D:\restic\logs\testlog.txt'
D:\restic\restic.exe --repo 'C:\restic_repo' --password-file 'C:\repopass' --verbose backup 'C:\Morrowind' > 'D:\restic\logs\testlog.txt'
& D:\restic\restic.exe --repo 'C:\restic_repo' --password-file 'C:\repopass' --verbose backup 'C:\Morrowind' 3>&1 2>> $ErrorLog | Tee-Object -Append $SuccessLog
```

Whether restic is run by a powershell script or in a live terminal doesn't seem to matter; the unwanted characters still end up in the output file.

I suspect normal redirection being affected as well as the two piped versions in my tests is because of this:
> Using the redirection operator with a file target is functionally equivalent to piping to Out-File with no extra parameters.
(https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_redirection?view=powershell-7.1)

It looks like you already have a very good idea of what's causing the behaviour, so I'm not sure how useful the below paragraphs will be...


That said, I was curious, so I did some digging, and think what I found might explain why the reports on which versions/terminals are affected appear to conflict at times in this thread.

I started with a git bisect between v0.11.0 and v0.12.0, which pointed to this commit (PR https://github.com/restic/restic/pull/3150) as where the problem first appeared for me (in powershell):

```
PS D:\restic> git bisect bad
b48f579530be5de1de08773a9ba27d241fcc9c1f is the first bad commit
```

Which is a little weird, given that particular PR was supposed to resolve this specific problem.

Bearing that in mind, I went backwards further, as v0.10.0 and v0.9.6 are mentioned at the beginning of this issue. I found I could reproduce the problem on 0.9.6 (again, still in powershell), but not on 0.10.0, which is the inverse of the original issue report.

Using the downloadable restic binaries, I was able to pin the first appearance of the problem down to between v0.8.3 and v0.9.0. Unfortunately, I was unable to bisect fully due to problems building some of the older restic commits (may be user error, not sure). I narrowed it down somewhat though:

```
PS D:\restic> git bisect skip
There are only 'skip'ped commits left to test.
The first bad commit could be any of:
38926d85760ab7ced41c088c94237db634088caf
a5c0cf23244c6da77434cc93a2eb3aa80c5dd1e4
9fac2ca832459801a730ad3cfcf24af2c23db22d
f279731168104727b1f47a5f9fcf460590f2792d
1af96fc6dd782176983386f8fa602552da9f5fbf
76b616451fcca84674f514d2bc41122f30bc4834
c703d21d5524835f8405f52b2d3ce6cee3e04df9
fd12a3af20c0fc7cd296c0f11f02d1f786b07538
0e78ac92d875e85c43aba9b315baca4be93d1383
We cannot bisect more!
```


Anyway, while bisecting, I checked again using cmd, and was surprised to find that the symptoms were the inverse of powershell; so when powershell works fine, cmd doesn't, and vice versa.

When using cmd, redirected output is normal for the 0.9.x series and 0.12.0, but has problems in versions v0.10.0 and v0.11.0. Powershell is the opposite, it has the problems with redirected output with the v0.9.x series and v0.12.0, but doesn't with v0.10.0 or v0.11.0. 
For the 0.8.x series of releases, both cmd and powershell appear to work fine.

Based on the above, I think the changes in PR 3105 and PR https://github.com/restic/restic/pull/2608 caused the problem to switch between powershell and cmd (and possibly other terminals?). This was apparently because of the isPipe condition being flipped, and is why there's some confusion over which terminals/verions are affected here.

tldr; redirected output looks like it has had this issue in powershell since v0.9.0. PR 2608 unintentionally fixed it for powershell but broke it for cmd (and possibly others?), then PR 3105 reverted this.

Hope the above helps!

**Comment by MichaelEischer** [comment]
Thanks for digging through previous restic version. The behavior in 0.10.0 and 0.11.0 is definitely wrong due to the broken isPipe condition.

restic tries to detect whether the output terminal is interactive. `canUpdateStatus` on windows (https://github.com/restic/restic/blob/8eb6a5805bdf290bee7fb4af1ace60a52a66197b/internal/ui/termstatus/terminal_windows.go#L85) currently checks whether stdout is a terminal or a pipe. In both cases restic assumes the output to be interactive, however with one major difference: when stdout is a pipe, then stdout is assumed to be connected to mintty which interprets posix escape codes. And the latter is what we see here in the log files.

So the way forward here is to fix the check whether stdout is connected to mintty and to ensure that it doesn't mistake normal output redirections for mintty. The question which remains here is: how to detect the presence of mintty?

**Comment by DRON-666** [comment]
>The question which remains here is: how to detect the presence of mintty?

You can check name of the output pipe.
See https://github.com/k-takata/ptycheck or http://git.661346.n2.nabble.com/PATCH-mingw-make-isatty-recognize-MSYS2-s-pseudo-terminals-dev-pty-td7654456.html

**Comment by DRON-666** [comment]
@MichaelEischer @fd0 I have the opposite question: has anyone tested this code in `mintty`? Right now I am running `restic backup` under `mintty` (MSYS2) and don't see any progress/status at all. I tested versions from 0.9.6 to 0.12.0 with the same results. Maybe this feature (mintty support) just never worked and can be removed?

**Comment by kmwoley** [comment]
If anyone is game for testing out a solution to this problem I put in a PR that appears to have resolved the issue for me (#3301).

**Comment by MichaelEischer** [comment]
I've opened #3325 which includes a much stricter detection of a mintty terminal. While implementing and testing that PR, I've noticed that #2706 (included in restic 0.12.0) prevented any interactive progress from being display on a mintty terminal. Which makes me wonder how low the mintty usage is.

**Comment by wendell77** [comment]
Unfortunately, v0.12.0 still displays escape characters in the cygwin environment. 
I am using it this way to remove the characters:
#restic .... | sed -e 's,\x1B\[[0-9;]*[a-zA-Z],,g' -e 's/[\x01-\x1F\x7F]//g'

**Comment by MichaelEischer** [comment]
For future reference, a log snippet using cygwin is shown here: https://forum.restic.net/t/output-remove-non-printing-characters/3812

@wendell77 The output in your forum post only contains non-interactive output. And for that restic 0.12.0 always prints the control characters independent of whether the output is a terminal or a file. Could you try whether PR #3325 solves the problem for you?

**Comment by wendell77** [comment]
@MichaelEischer The binary I'm using is restic 0.12.0 (2021-02-14). Where can I find a version compiled with this modification?
